programs/leap-util/actions/snapshot.cpp

```diff
@@ -10,29 +16,86 @@
-   sub->add_option("--input-file,-i", opt->input_file, "snapshot file to convert to json format, writes to <file>.json if output file not specified (tmp state dir used), and exit.")->required();
-   sub->add_option("--output-file,-o", opt->output_file, "the file to write the output to (absolute or relative path).  if not specified then output is to stdout.");
+   // subcommand -convert snapshot to json
+   auto to_json = sub->add_subcommand("to-json", "Convert snapshot file to json format");
+   to_json->add_option("--input-file,-i", opt->input_file, "Snapshot file to convert to json format, writes to <file>.json if output file not specified (tmp state dir used).")->required();
+   to_json->add_option("--output-file,-o", opt->output_file, "The file to write the output to (absolute or relative path).  If not specified then output is to <input-file>.json.");
 
-   sub->callback([this]() {
-      int rc = run_subcommand();
-      // properly return err code in main
-      if(rc) throw(CLI::RuntimeError(rc));
+   to_json->callback([this]() {
+      try {
+         int rc = run_subcommand();
+         if(rc) throw(CLI::RuntimeError(rc));
+      } catch(...) {
+         print_exception();
+         throw(CLI::RuntimeError(-1));
+      }
    });
 }
 
 int snapshot_actions::run_subcommand() {
    if(!opt->input_file.empty()) {
       if(!fc::exists(opt->input_file)) {
-         std::cerr << "cannot load snapshot, " << opt->input_file << " does not exist" << std::endl;
+         std::cerr << "cannot load snapshot, " << opt->input_file
+                   << " does not exist" << std::endl;
          return -1;
       }
    }
-   // todo: implementation here
 
+   bfs::path snapshot_path = opt->input_file;
+   bfs::path json_path = opt->output_file.empty()
+                               ? snapshot_path.generic_string() + ".json"
+                               : opt->output_file;
+   // pull chain id
+   auto infile = std::ifstream(snapshot_path.generic_string(),
+                               (std::ios::in | std::ios::binary));
+   istream_snapshot_reader reader(infile);
+   reader.validate();
+   auto chain_id = controller::extract_chain_id(reader);
+   infile.close();
+
+   // setup controller
+   bfs::path temp_dir = boost::filesystem::temp_directory_path() /
+                        boost::filesystem::unique_path();
+   bfs::path state_dir = temp_dir / "state";
+   bfs::path blocks_dir = temp_dir / "blocks";
+   std::unique_ptr<controller> control;
+   protocol_feature_set pfs;
+   controller::config cfg;
+   cfg.blocks_dir = blocks_dir;
+   cfg.state_dir = state_dir;
+
+   try {
+      auto infile = std::ifstream(snapshot_path.generic_string(),
+                                  (std::ios::in | std::ios::binary));
+      auto reader = std::make_shared<istream_snapshot_reader>(infile);
+      control.reset(new controller(cfg, std::move(pfs), chain_id));
+      control->add_indices();
+      control->startup([]() {}, []() { return false; }, reader);
```

Comment:
The first param expects a `shutdown()` lambda, providing it an empty lambda does not handle errors correctly.
For example:

```
➜  22-11-07 15:13:46~/ext/leap/cmake-build-relwithdebinfo/bin git:(add-snapshot-subcommand) ✗ ./leap-util snapshot to-json -i snapshot-2022-09-17-11-wax-v6-0203904304.bin
info  2022-11-07T21:14:28.648 leap-util platform_timer_accurac:62     compute_and_print_ti ] Checktime timer accuracy: min:0us max:37us mean:3us stddev:3us
info  2022-11-07T21:14:28.648 leap-util controller.cpp:564            startup              ] Starting initialization from snapshot, this may take a significant amount of time
error 2022-11-07T21:14:36.866 leap-util controller.cpp:581            startup              ] Failed initialization from snapshot - db storage not configured to have enough storage for the provided snapshot, please increase and retry snapshot
info  2022-11-07T21:14:36.866 leap-util snapshot.cpp:85               run_subcommand       ] Writing snapshot: snapshot-2022-09-17-11-wax-v6-0203904304.bin.json
debug 2022-11-07T21:14:36.925 leap-util controller.cpp:2436           clear_expired_input_ ] removed 0 expired transactions of the 0 input dedup list, pending block time 2022-09-17T11:00:05.500
info  2022-11-07T21:14:49.016 leap-util snapshot.cpp:99               run_subcommand       ] Completed writing snapshot: snapshot-2022-09-17-11-wax-v6-0203904304.bin.json
➜  22-11-07 15:14:49~/ext/leap/cmake-build-relwithdebinfo/bin git:(add-snapshot-subcommand) ✗ echo $?
0
```

Reply:
Fixed, now return code indicates failure